stdlib/public/core/DiscontiguousSlice.swift

```diff
@@ -0,0 +1,450 @@
+      i.base == subranges.ranges[i._rangeOffset].lowerBound {
+      // Go to next subrange
+      let nextOffset = i._rangeOffset - 1
+      let nextRange = subranges.ranges[nextOffset]
+      let nextBase = base[nextRange].index(before: nextRange.upperBound)
+      return Index(_rangeOffset: nextOffset, base: nextBase)
+    } else {
+      // Move within current subrange
+      let currentRange = subranges.ranges[i._rangeOffset]
+      let nextBase = base[currentRange].index(before: i.base)
+      return Index(_rangeOffset: i._rangeOffset, base: nextBase)
+    }
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice where Base: MutableCollection {
+  /// Accesses the element at the specified position.
+  ///
+  /// For example, you can replace an element of an array by using its
+  /// subscript.
+  ///
+  ///     var streets = ["Adams", "Bryant", "Channing", "Douglas", "Evarts"]
+  ///     streets[1] = "Butler"
+  ///     print(streets[1])
+  ///     // Prints "Butler"
+  ///
+  /// You can subscript a collection with any valid index other than the
+  /// collection's end index. The end index refers to the position one
+  /// past the last element of a collection, so it doesn't correspond with an
+  /// element.
+  ///
+  /// - Parameter position: The position of the element to access. `position`
+  ///   must be a valid index of the collection that is not equal to the
+  ///   `endIndex` property.
+  ///
+  /// - Complexity: O(1)
+  public subscript(i: Index) -> Base.Element {
+    get {
+      _base[subranges.ranges[i._rangeOffset]][i.base]
+    }
+    set {
+      _base[subranges.ranges[i._rangeOffset]][i.base] = newValue
+    }
+  }
+}
+
+// MARK: Accessing DiscontiguousSlices
+
+extension Collection {
+  /// Accesses a view of this collection with the elements at the given
+  /// indices.
+  ///
+  /// - Parameter subranges: The indices of the elements to retrieve from this
+  ///   collection.
+  /// - Returns: A collection of the elements at the positions in `subranges`.
+  ///
+  /// - Complexity: O(1)
+  @available(SwiftStdlib 5.11, *)
+  public subscript(subranges: RangeSet<Index>) -> DiscontiguousSlice<Self> {
+    DiscontiguousSlice(_base: self, subranges: subranges)
+  }
+}
+
+extension MutableCollection {
+  /// Accesses a mutable view of this collection with the elements at the
+  /// given indices.
+  ///
+  /// - Parameter subranges: The ranges of the elements to retrieve from this
+  ///   collection.
+  /// - Returns: A collection of the elements at the positions in `subranges`.
+  ///
+  /// - Complexity: O(1) to access the elements, O(*m*) to mutate the
+  ///   elements at the positions in `subranges`, where *m* is the number of
+  ///   elements indicated by `subranges`.
+  @available(SwiftStdlib 5.11, *)
+  public subscript(subranges: RangeSet<Index>) -> DiscontiguousSlice<Self> {
+    get {
+      DiscontiguousSlice(_base: self, subranges: subranges)
+    }
+    set {
```

Comment:
API note:

It is weirdly over-constraining that we only allow assigning instances of `DiscontiguousSlice` here.

It would make this far more useful if we generalized this to a function that takes an arbitrary collection:

```swift
extension MutableCollection {
  public mutating func updateSubranges(
    _ subranges: RangeSet<Index>, 
    with newElements: some Collection<Element>
  ) {
    // Pretty much the same code than what we currently have in this setter...
  }
}
```

We can either keep the subscript setter (if we are confident that its semantics are obvious), or remove it altogether.

Additionally, it doesn't feel right to keep a generic algorithm like this non-inlinable -- it expressed entirely through public API. (However, if we regret this choice, we can always add `@inlinable` later.)

Reply:
It's worth thinking about how this could/would be used, especially given the removal of `MutableCollection` conformance for `DiscontiguousSlice` (which means there aren't any mutating methods to use with chaining after the subscript). To use the subscript setter here, you'd need to:

1. Retrieve a discontiguous slice from a collection
2. Retrieve an index from within that slice
3. Modify the elements in the discontiguous slice with the index (this creates a copy of the entire base collection)
4. Assign the discontiguous back to the collection

Given the required copy of the base collection, this is almost certainly a worse experience with worse performance than making them copy the contents of the slice into an array, or just using the index they found in (2) to modify the original collection.

Reply:
I agree having an `updateSubranges` that takes an arbitrary collection would be better here, but I also think this is an issue with `Slice` as well and perhaps might be better off discussed in a separate proposal which can improve the `MutableCollection` + slicing APIs for both `Slice` and `DiscontiguousSlice`?

As for this subscript, the only other way it could be used would be to assign one slice to another via this subscript, for example:

```swift
let numbers = [1, 2, 3, 4]
let evenIndices = numbers.indices { $0 % 2 == 0 }
let oddIndices = numbers.indices { $0 % 2 != 0 }
numbers[evenIndices] = numbers[oddIndices]
print(numbers) // [1, 1, 3, 3]
```

I don't think this is a very common use case given the constraint that the right hand side must be a `DiscontiguousSlice` of the same collection and of the same length, so I'd be ok removing this for now if we deem the semantics of it not quite right so that we can come back and add something like `updateSubrange` + `updateSubranges` later

Reply:
I've addressed this by removing this setter since I do agree that the semantics aren't obvious and it can likely use a rethink with `Slice` included as a whole